icu4c/source/common/locid.cpp

```diff
@@ -326,12 +326,109 @@ struct Locale::Heap : public UMemory {
         }
     }
 
-    // Move should be done on the std::unique_ptr object that owns this.
+    // Move should be done on the owner of the pointer to this object.
     Heap(Heap&&) noexcept = delete;
 
     ~Heap() = default;
 };
 
+Locale::Payload::HeapPtr& Locale::Payload::HeapPtr::operator=(const Heap& other) {
+    U_ASSERT(type == eBOGUS);
+    UErrorCode status = U_ZERO_ERROR;
+    ptr = new Heap(other, status);
+    if (ptr == nullptr || U_FAILURE(status)) {
+        delete ptr;
+    } else {
+        type = eHEAP;
+    }
+    return *this;
+}
+
+Locale::Payload::HeapPtr& Locale::Payload::HeapPtr::operator=(HeapPtr&& other) noexcept {
+    U_ASSERT(type == eBOGUS);
+    ptr = other.ptr;
+    type = eHEAP;
+    other.type = eBOGUS;
+    return *this;
+}
+
+template <typename Fn0, typename Fn1, typename Fn2, typename... Args>
+auto Locale::Payload::visit(Fn0 fn0, Fn1 fn1, Fn2 fn2, Args... args) const {
+    switch (stat.type) {
+        case eBOGUS:
+            return fn0(args...);
+        case eNEST:
+            return fn1(nest, args...);
+        case eHEAP:
+            return fn2(*heap.ptr, args...);
+        default:
+            UPRV_UNREACHABLE_EXIT;
+    };
+}
+
+void Locale::Payload::copy(const Payload& other) {
+    other.visit([](Payload*) {},
+                [](const Nest& nest, Payload* dst) { dst->nest = nest; },
+                [](const Heap& heap, Payload* dst) { dst->heap = heap; }, this);
+}
+
+void Locale::Payload::move(Payload&& other) noexcept {
+    other.visit(
+        [](Payload*, Payload*) {},
+        [](const Nest& nest, Payload* dst, Payload*) { dst->nest = nest; },
+        [](const Heap&, Payload* dst, Payload* src) { dst->heap = std::move(src->heap); }, this, &other);
+}
+
+Locale::Payload::~Payload() {
+    if (stat.type == eHEAP) { delete heap.ptr; }
+}
+
+Locale::Payload::Payload(const Payload& other) : stat{eBOGUS} { copy(other); }
+Locale::Payload::Payload(Payload&& other) noexcept : stat{eBOGUS} { move(std::move(other)); }
+
+Locale::Payload& Locale::Payload::operator=(const Payload& other) {
+    if (this != &other) {
+        setToBogus();
+        copy(other);
+    }
+    return *this;
+}
+
+Locale::Payload& Locale::Payload::operator=(Payload&& other) noexcept {
+    if (this != &other) {
+        setToBogus();
+        move(std::move(other));
+    }
+    return *this;
+}
+
+void Locale::Payload::setToBogus() {
+    this->~Payload();
+    stat.type = eBOGUS;
+}
+
+LocalPointer<Locale::Heap> Locale::Payload::release() {
+    U_ASSERT(stat.type == eHEAP);
+    stat.type = eBOGUS;
+    return LocalPointer<Heap>{heap.ptr};
+}
+
+template <> Locale::Nest& Locale::Payload::emplace() {
+    this->~Payload();
+    ::new(&nest) Nest();
+    return nest;
+}
+
+template <typename T, typename... Args> T& Locale::Payload::emplace(Args&&... args) {
```

Comment:
If the return type is always `Heap*` can we make that explicit and remove tparam T?

How about calling this function emplaceHeap() for clarity?